### PR TITLE
Report shadowed type parameters in generic typedefs

### PR DESCRIPTION
**Pull request by srawlins**
Partial fix for #1027 but I suspect not the whole story...

### CODE REVIEW

File: lib/src/rules/avoid_shadowing_type_parameters.dart [hunk 1]
```diff
@@ -59,6 +60,11 @@ class _Visitor extends SimpleAstVisitor<void> {
     _checkAncestorParameters(functionExpression.typeParameters, node);
   }
 
+  @override
+  void visitGenericTypeAlias(GenericTypeAlias node) {
+    _checkForShadowing(node.functionType.typeParameters, node.typeParameters);
```
Comment (bwilkerson) [inline lib/src/rules/avoid_shadowing_type_parameters.dart]: `_checkForShadowing` doesn't check the first parameter to ensure that it isn't `null`. We should either change that (and clean up existing code) or add a null check here.

File: test/rules/avoid_shadowing_type_parameters.dart [hunk 2]
```diff
@@ -51,3 +51,7 @@ class E {
 
   void fn4<T>() {} // OK
 }
+
+typedef Fn1<T> = void Function<T>(T); // LINT
+typedef Fn2<T> = void Function<U>(T); // OK
+typedef Fn3<T> = void Function<U>(U); // OK
```
Comment (bwilkerson) [inline test/rules/avoid_shadowing_type_parameters.dart]: Let's add two more test cases:
```
typedef Fn4<T> = void Function(T); // OK
typedef Fn5 = void Function<T>(T); // OK
```

File: lib/src/rules/avoid_shadowing_type_parameters.dart (hunk 1 above)
Comment (srawlins) [reply lib/src/rules/avoid_shadowing_type_parameters.dart]: Ah, good catch. I think I like null-checking before the function here. Done.

File: test/rules/avoid_shadowing_type_parameters.dart (hunk 2 above)
Comment (srawlins) [reply test/rules/avoid_shadowing_type_parameters.dart]: Done, thanks.

**Pull request merged**